torch/csrc/jit/pickler.cpp

```diff
@@ -45,15 +51,81 @@ const std::vector<char>& Pickler::stack() {
 
 void Pickler::start() {
   push<OpCode>(OpCode::PROTO);
-  push<uint8_t>(2);
+  push<uint8_t>(PROTOCOL_VERSION);
+}
 
-  // All attributes get pushed into a list and their indices saved in the
+void Pickler::startTuple() {
+  // All attributes get pushed into a tuple and their indices saved in the
   // module def
   push<OpCode>(OpCode::MARK);
 }
 
-void Pickler::finish() {
+void Pickler::endTuple() {
   push<OpCode>(OpCode::TUPLE);
+}
+
+void Pickler::finish() {
+  push<OpCode>(OpCode::STOP);
+
+
+  // Add the binary data for all the tensors to be included in the same binary
+  // TODO: The pickler should be refactored to stream out to a stream directly
+  // instead of staging in the stack_ array
+  if (literal_tensors_.size() > 0) {
+    // As another pickle program in the same binary archive, add a list of
+    // keys for each tensor (see torch/serialization.py)
+    start();
+    push<OpCode>(OpCode::MARK);
```

Comment:
Is this right? getTensorKey will not be unique per tensor. Only per storage.

Reply:
That is right, it should be per storage (the deserialization uses these to skip loading the same storage twice)

https://github.com/pytorch/pytorch/blob/440aac082a804d88f9bbaa5f5f8dfec8f11936a4/torch/serialization.py#L576-L582